### PR TITLE
Add ability to draw mesh edges and add more drawing modes

### PR DESCRIPTION
**Pull request by yanchith**
This adds 4 drawing modes and the ability to switch between them:

- Shaded:          What we had until now
- Edges:           Draws just the edges of the mesh
- ShadedEdges:     Draws both the shaded view and the visible edges
- ShadedEdgesXray: Draws both the shaded view and all the edges, even
                   occluded

The edge drawing is done with a shader trick, which seemed like the
simplest solution initially (required the least re-architecting), but
it might not be enough for our later needs. See `viewport.frag` for
more.

The scene renderer is refactored internally to support
(very trivial and mostly incorrect) rendering of transparent objects:
`Edges` and `ShadedEdgesXray` use the new transparent render pipeline.

Both rendering of faces and edges is done by the same shader, with
flags used to enable rendering of one or the other, or both (for
`ShadedEdges`). `ShadedEdgesXray` is rendered in two passes - first
`Shaded`, then `Edges`.

**Reviewers:**

- Please also test visually on your machines (this paint job was done on an aging macOS MBP).
- Also, I silenced ~the lamb~ `clippy::too_many_arguments` on two occasions. I don't believe the lint is wrong, but GPU programming with `wgpu` is a bit verbose and listening to clippy would IMO make the code less readable in this case.

### CODE REVIEW

File: src/renderer/scene_renderer.rs [hunk 1]
```diff
@@ -719,6 +876,115 @@ fn apply_wgpu_correction_matrix(projection_matrix: &Matrix4<f32>) -> Matrix4<f32
     wgpu_correction_matrix * projection_matrix
 }
 
+fn barycentric_sequence_iter() -> impl Iterator<Item = u32> {
```
Comment (ondrowan) [inline src/renderer/scene_renderer.rs:880]: This could use a docstring.

File: src/renderer/mod.rs [hunk 2]
```diff
@@ -1,12 +1,12 @@
...
 use std::fmt;
 
 use nalgebra::base::Matrix4;
 
 use self::imgui_renderer::{ImguiRenderer, ImguiRendererOptions};
 use self::scene_renderer::{
-    SceneRenderer, SceneRendererAddGeometryError, SceneRendererClearFlags, SceneRendererOptions,
+    SceneRenderer, SceneRendererAddGeometryError, SceneRendererClearFlags,
+    SceneRendererDrawGeometryMode, SceneRendererGeometry, SceneRendererGeometryId,
+    SceneRendererOptions,
```
Comment (ondrowan) [inline src/renderer/mod.rs]: Not fully part of this PR and I should've pointed it out earlier, but I'm not a fan of this kind of naming. Besides `SceneRenderer` and maybe errors, none of these should contain `SceneRenderer` prefix. It should be obvious it comes from there and if it's not, users can alias it.

**Review by ondrowan — CHANGES_REQUESTED** [review]
I've tried it on Windows machine with both vulkan and D3D12 and it seems to work well.

Unrelated to this PR, I've found out that D3D12 rendering makes the controls quite unresponsive and it seems the higher value for MSAA I choose, the worse it is. It shows I have good framerate, but if I try to move camera it responds with a lag. If MSAA is 1, it happens only for a second of two after starting the app. If it's higher, it's unusable. It happens only in debug mode though, so might not be a cause for panic.

File: src/renderer/mod.rs (hunk 2 above)
Comment (yanchith) [reply src/renderer/mod.rs]: Hmm there is some merit to what you say. I will ponder it.

Comment (yanchith) [reply src/renderer/mod.rs]: Just looked elsewhere in the codebase, errors do not share the prefix, so I'll remove it from them as well.

**Comment by yanchith** [comment]
> Unrelated to this PR, I've found out that D3D12 rendering makes the controls quite unresponsive and it seems the higher value for MSAA I choose, the worse it is.

I think we'll need to have a pass on dealing with the winit/wgpu bugs at some point, documenting all of them, trying to update and discerning what changed. Worst case we might need to get our hands dirty and fix them upstream.

**Review by ondrowan — APPROVED** [review]

**Review by janper — APPROVED** [review]
great! 
image

**Pull request merged**